### PR TITLE
htlcswitch & lnwallet: Route packet forwards by (chan ID, HTLC ID)

### PR DESCRIPTION
**Pull request by jimpo**
This changes how the switch routes forwarding packets. In the switch, link, and channel currently, payment hash is used an an identifier for an HTLC, which is not great because the protocol supports multiple HTLCs with the same payment hash. This appears to not be handled perfectly as is (eg. see [here](https://github.com/lightningnetwork/lnd/blob/master/lnwallet/channel.go#L3107)). Instead, it is better to identify HTLCs by the tuple of channel and unique HTLC ID (as specified in the update add message).

This has the benefit that the circuit map can definitively map every offered HTLC resulting from a forward back to exactly one received HTLC which is important in terms of accountability (verifying that we haven't offered HTLCs without a promise of funds or duplicating outbound HTLCs). I think ideally the circuit map would be extended to include pending payments originating from the local node so that *every* outbound HTLC is linked back one-to-one with its origin.

Status: integration test is broken.

### CODE REVIEW

File: lnwire/short_channel_id.go [hunk 1]
```diff
@@ -37,3 +42,11 @@ func (c *ShortChannelID) ToUint64() uint64 {
 	return ((uint64(c.BlockHeight) << 40) | (uint64(c.TxIndex) << 16) |
 		(uint64(c.TxPosition)))
 }
+
+// String generates a string representation of the channel ID. This returns the
```
Comment (Roasbeef) [inline lnwire/short_channel_id.go]: Returning this encoding doesn't really convey any meaning when displayed. I've thought a bit about how best to display this (instead of just the integer everywhere), how about something along the lines of:
```
height:index:pos
```
or
```
height/index/pos
```

File: htlcswitch/circuit.go [hunk 2]
```diff
@@ -1,135 +1,169 @@
...
-	e ErrorEncrypter) *paymentCircuit {
-
-	return &paymentCircuit{
-		Src:            src,
-		Dest:           dest,
-		PaymentHash:    key,
-		RefCount:       1,
-		ErrorEncrypter: e,
-	}
+// circuitKey is a channel ID, hash ID tuple used as an identifying key for a
```
Comment (Roasbeef) [inline htlcswitch/circuit.go]: hash -> htlc

File: htlcswitch/circuit.go [hunk 3]
```diff
@@ -1,135 +1,169 @@
...
+// payment circuit.
+type circuitKey struct {
```
Comment (Roasbeef) [inline htlcswitch/circuit.go]: Not clear from this `godoc` comment if this is the channel id + htlc id or the _incoming_ or _outgoing_ htlc. From reading the diff a bit more, I see it's for the dest. I recommend adding a `dest*` prefix to the fields.

File: htlcswitch/link.go [hunk 4]
```diff
@@ -610,6 +612,20 @@ func (l *channelLink) handleDownStreamPkt(pkt *htlcPacket, isReProcess bool) {
 			"local_log_index=%v, batch_size=%v",
 			htlc.PaymentHash[:], index, l.batchCounter+1)
 
+		// If packet was forwarded from another channel link then we should
```
Comment (Roasbeef) [inline htlcswitch/link.go]: I think this should instead be done once the HTLC has actually been _locked into_ the commitment transaction. This isn't absolutely necessary now, but seems like a requirement for your switch persistent ideas to be implemented properly. In order to do this, we'd need to modify the `ReceiveRevocation` method to also scan _our_ update log and return any locked-in entries. Doing this would be independently useful as it would add another synchronization hook for many unit tests in the project that currently jut sleep as they wait for an HTLC to be locked in.

File: htlcswitch/switch.go [hunk 5]
```diff
@@ -562,37 +532,48 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
...
+					htlc.Reason)
+			}
+
+			packet.dest = circuit.SrcChanID
+			packet.destID = circuit.SrcHTLCID
 		}
 
-		// Propagating settle/fail htlc back to src of add htlc packet.
-		source, err := s.getLinkByShortID(circuit.Src)
+		source, err := s.getLinkByShortID(packet.dest)
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Shouldn't this be looking up the _source_ of the packet?

File: htlcswitch/switch.go [hunk 6]
```diff
@@ -562,37 +532,48 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
...
+
+			// Obfuscate the error message for fail updates before sending back
+			// through the circuit.
+			if htlc, ok := htlc.(*lnwire.UpdateFailHTLC); ok && !packet.isObfuscated {
+				htlc.Reason = circuit.ErrorEncrypter.IntermediateEncrypt(
+					htlc.Reason)
+			}
+
+			packet.dest = circuit.SrcChanID
+			packet.destID = circuit.SrcHTLCID
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: This is a bit hacky, why are we unable to just _always_ use the source below? (as we did before)

File: htlcswitch/switch_test.go [hunk 7]
```diff
@@ -66,6 +67,15 @@ func TestSwitchForward(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	s.addCircuit(&PaymentCircuit{
```
Comment (Roasbeef) [inline htlcswitch/switch_test.go]: Why is necessary to _manually_ add the payment circuit? It seems that the link the HTLC is being forwarded to should be populating it once it handles the downstream message.

File: htlcswitch/link.go [hunk 8]
```diff
@@ -659,7 +658,7 @@ func (l *channelLink) handleDownStreamPkt(pkt *htlcPacket, isReProcess bool) {
 	case *lnwire.UpdateFailHTLC:
 		// An HTLC cancellation has been triggered somewhere upstream,
 		// we'll remove then HTLC from our local state machine.
-		logIndex, err := l.channel.FailHTLC(pkt.payHash)
+		err := l.channel.FailHTLC(pkt.destID)
```
Comment (Roasbeef) [inline htlcswitch/link.go]: I find the src/dest breakdown a bit confusing here. In my mind, this should be the src ID as a previously _outgoing_ HTLC (from the PoV of this channel) has now been cancelled by the dest link (where we extended the forwarded HTLC). 

What do you think of switching to incoming/outgoing?

File: htlcswitch/link.go (hunk 4 above)
Comment (jimpo) [reply htlcswitch/link.go]: I agree that is a better time to add to the circuit map. Another complication is that we'll need to add `sourceChannelID` and `sourceHTLCID` to PaymentDescriptor so that we know what to put in the circuit map after they revoke. That is part of my master plan in terms of persistence and stuff though, so I'll go ahead with it and we can see how it looks.

File: htlcswitch/link.go (hunk 8 above)
Comment (jimpo) [reply htlcswitch/link.go]: Yeah, I agree the incoming/outgoing terminology is less ambiguous. Will change.

File: htlcswitch/switch.go (hunk 6 above)
Comment (jimpo) [reply htlcswitch/switch.go]: I may have reversed some of the terminology because I was thinking in terms of `src` and `destination` packets, not HTLCs. Like the destination of an `update_add` is the opposite of the destination for the `update_{fail,settle}` going the the other way. Hopefully, incoming/outgoing HTLC terminology clears things up.

File: htlcswitch/switch_test.go (hunk 7 above)
Comment (jimpo) [reply htlcswitch/switch_test.go]: Because we are interacting with mock links, instead of the real links which are responsible for adding to the circuit map.

**Comment by Roasbeef** [comment]
When you get a chance, can you rebase this? The latest push to implement channel retransmission resulted in a bit of a overhaul within the link and switch. Thanks!

File: htlcswitch/circuit.go [hunk 9]
```diff
@@ -1,135 +1,170 @@
...
 	}
+	circuit := cm.circuits[key]
+
+	cm.mtx.RUnlock()
+	return circuit
+}
+
+// LookupByPaymentHash looks up and returns any payment circuits with a given
+// payment hash.
+func (cm *CircuitMap) LookupByPaymentHash(hash [32]byte) []*PaymentCircuit {
```
[the remark this answers is not in the document]
Comment (jimpo) [reply htlcswitch/circuit.go]: Yes, it would be used when pipelining fulfills. As soon as we get a `update_fulfill` on an outgoing channel, we should immediately turn around and fulfill on all incoming channels with the same payment hash, even before the first fulfill commits.

File: htlcswitch/circuit.go [hunk 10]
```diff
@@ -1,135 +1,170 @@
...
+	delete(cm.circuits, key)
 
-	if circuit, ok := m.circuits[key]; ok {
-		if circuit.RefCount--; circuit.RefCount == 0 {
-			delete(m.circuits, key)
+	// Remove key from hash index.
+	index := -1
+	circuitsWithHash := cm.hashIndex[circuit.PaymentHash]
+	for i, indexedCircuit := range circuitsWithHash {
+		if indexedCircuit == circuit {
```
[the remark this answers is not in the document]
Comment (jimpo) [reply htlcswitch/circuit.go]: I'll do it if you really want, but I'm not convinced it'll be better for this use case. Firstly, LookupByPaymentHash has to return a slice. Second, I expect most of the time payment hashes will not be reused and there will be one circuit per payment hash, or at least a very small number.

[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/circuit.go]: Well, it returns a slice currently but doesn't have to necessarily. In fact, I think a slice should be derived from the map during every call to `LookupByPaymentHash`. I just noticed that we return the slices directly, whose contents be mutated by subsequent calls to `Remove`, which does an in-place swap/truncate. Using a map will also ensure that `Add` and `Remove` are idempotent, this implementation would break if they are not called the same number of times.

[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/circuit.go]: I think as is you could cause a nil pointer deference by
```
circuits := cm.LookupByPaymentHash(hash)
cm.Remove(circuits[0])
for i, circuit := range circuits {
    // Will fail at i = len(circuits) - 1
    circuit.variable
}
```

File: htlcswitch/circuit.go [hunk 11]
```diff
@@ -1,135 +1,172 @@
...
+	if !ok {
+		return errors.Errorf("Can't find circuit in hash index for HTLC %v",
+			key)
+	} else if _, ok = circuitsWithHash[circuit]; !ok {
+		return errors.Errorf("Can't find circuit in hash index for HTLC %v",
+			key)
+	} else if len(circuitsWithHash) == 1 {
+		delete(cm.hashIndex, circuit.PaymentHash)
+	} else {
+		delete(circuitsWithHash, circuit)
```
[the remark this answers is not in the document]
Comment (jimpo) [reply htlcswitch/circuit.go]: But shouldn't the circuit get garbage collected along with the map when the map is deleted from the hash index?

File: htlcswitch/switch.go [hunk 12]
```diff
@@ -566,37 +559,58 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
...
-		// Exit if we can't find and remove the active circuit to
-		// continue propagating the fail over.
-		circuit, err := s.circuits.remove(packet.payHash)
-		if err != nil {
-			err := errors.Errorf("unable to remove "+
-				"circuit for payment hash: %v", packet.payHash)
-			log.Error(err)
-			return err
-		}
+		if packet.incomingChanID == (lnwire.ShortChannelID{}) {
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Could use a comment here to indicate that the `packet.incomingChanID == (lnwire.ShortChannelID{})` fragment indicates the HTLC was initiated from a local subsystem. Alternatively, could be extracted out to a helper method on the packet as it reoccurs in a few places.

Comment (Roasbeef) [reply htlcswitch/switch.go]: It would seem now that the circuit is _only_ removed if the payment was initiated locally? If so this is a change in behavior and incorrect, as now the circuit map will never properly be garbage collected.

File: htlcswitch/switch.go [hunk 13]
```diff
@@ -566,37 +559,58 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
...
+			err := s.circuits.Remove(packet.outgoingChanID,
+				packet.outgoingHTLCID)
+			if err != nil {
+				log.Warnf("Failed to close completed onion circuit: "+
+					"(%s, %d) <-> (%s, %d)",
+					circuit.IncomingChanID, circuit.IncomingHTLCID,
+					circuit.OutgoingChanID, circuit.OutgoingHTLCID)
+			} else {
+				log.Debugf("Closed completed onion circuit: "+
+					"(%s, %d) <-> (%s, %d)",
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: The payment hash should also be logged as before. The hash is a global identifier and gives more info that just the chanIDs+htlcIDs.

File: htlcswitch/link.go [hunk 14]
```diff
@@ -652,55 +652,48 @@ func (l *channelLink) handleDownStreamPkt(pkt *htlcPacket, isReProcess bool) {
 			// machine, as a result, we'll signal the switch to
 			// cancel the pending payment.
 			default:
+				log.Infof("Unable to handle downstream add HTLC: %v", err)
```
Comment (Roasbeef) [inline htlcswitch/link.go]: Should be at `Errorf` logging (or maybe just `Warnf`).

File: htlcswitch/link.go [hunk 15]
```diff
@@ -652,55 +652,48 @@ func (l *channelLink) handleDownStreamPkt(pkt *htlcPacket, isReProcess bool) {
...
+
 				var (
-					isObfuscated bool
+					localFailure = false
```
Comment (Roasbeef) [inline htlcswitch/link.go]: Why the change in naming? As a result, the semantics have now been flipped everywhere.

File: htlcswitch/switch.go [hunk 16]
```diff
@@ -406,30 +424,46 @@ func (s *Switch) handleLocalDispatch(payment *pendingPayment, packet *htlcPacket
...
-				payment.paymentHash[:], err)
-			log.Error(userErr)
-			payment.err <- &ForwardingError{
+		var failure *ForwardingError
+		if packet.localFailure {
+			var userErr string
+			r := bytes.NewReader(htlc.Reason)
+			failureMsg, err := lnwire.DecodeFailure(r, 0)
+			if err != nil {
+				userErr = fmt.Sprintf("unable to de-obfuscate onion failure, "+
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Error message is now incorrect. It should state it's unable to _decode_ the wire error.

File: htlcswitch/switch.go [hunk 17]
```diff
@@ -566,37 +576,60 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
...
 	case *lnwire.UpdateFufillHTLC, *lnwire.UpdateFailHTLC:
-		// Exit if we can't find and remove the active circuit to
-		// continue propagating the fail over.
-		circuit, err := s.circuits.remove(packet.payHash)
-		if err != nil {
-			err := errors.Errorf("unable to remove "+
-				"circuit for payment hash: %v", packet.payHash)
-			log.Error(err)
-			return err
+		if !packet.isRouted {
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Ahh was commenting on an out of date diff.

Comment (Roasbeef) [reply htlcswitch/switch.go]: Would be worth adding a comment here stating that things are only routed once they've reached the _outgoing_ channel. So in the case that we initiate a local payment, then there's nothing in the circuit map to delete.

File: htlcswitch/link.go (hunk 15 above)
Comment (jimpo) [reply htlcswitch/link.go]: It represents something different from what `isObfuscated` used to.

File: htlcswitch/switch.go (hunk 17 above)
Comment (jimpo) [reply htlcswitch/switch.go]: That's not what `isRouted` refers to. `isRouted` means the incomingChanID and incomingHTLCID are already set and do not need to be looked up in the circuit map. Local payments still get added and removed from the circuit map.

File: htlcswitch/link.go [hunk 18]
```diff
@@ -755,6 +756,28 @@ func (l *channelLink) handleDownStreamPkt(pkt *htlcPacket, isReProcess bool) {
 		htlc.ChanID = l.ChanID()
 		htlc.ID = pkt.incomingHTLCID
 
+		// Finally, we send the HTLC message to the peer which
```
Comment (Roasbeef) [inline htlcswitch/link.go]: I think this commit should be removed all together. This doesn't add anything but additional overhead. If we want to communicate directly to the peer, then we should just to so. There's no need to create this circular loop.

Comment (jimpo) [reply htlcswitch/link.go]: Removed.
